### PR TITLE
[FIX] Distances: handling errors due to too large arrays

### PR DESCRIPTION
**Pull request by jerneju**
##### Issue
https://sentry.io/biolab/orange3/issues/196744070/events/5529227746/

##### Description of changes
Users sees an error message when calculating too large arrays and Orange does not crash.

##### Includes
- [X] Code changes
- [X] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by jerneju** [comment]
https://sentry.io/biolab/orange3/issues/206013331/

**Comment by janezd** [comment]
@astaric, *similar code warnings* are stupid. See this -- it codeclimate complains about two totally different lists of error in `class Error`.

File: Orange/widgets/unsupervised/owdistances.py [hunk 1]
```diff
@@ -48,6 +48,8 @@ class Error(OWWidget.Error):
         dense_metric_sparse_data = Msg("Selected metric does not support sparse data")
         empty_data = Msg("Empty data set")
         mahalanobis_error = Msg("{}")
+        distances_memory_error = Msg("Not enough memory.")
+        distances_value_error = Msg("Error occurred while calculating distances: {}")
```
Comment (janezd) [inline Orange/widgets/unsupervised/owdistances.py]: The error can be long, so `"Error occurred while calculating distances\n{}"`. Skip the colon, so the error looks OK, and the tooltip shows the entire message.

File: Orange/widgets/unsupervised/owdistances.py [hunk 2]
```diff
@@ -144,7 +157,7 @@ def compute_distances(self, metric, data):
                 self.Error.mahalanobis_error(e)
                 return
 
-        return metric(data, data, 1 - self.axis, impute=True)
+        return compute(metric, data)
```
Comment (janezd) [inline Orange/widgets/unsupervised/owdistances.py]: A stylistic comment. I don't like the obvious radon fixes --- here's a code that's too complex so let's tear out a chunk of it into a separate function.

The main job of `compute_distances` is not to run various checks and fixes, and in the end, yes, call a local function that computes distances.

Turn it around. The "auxiliary" code should be the code that prepares the data and not the code that does the work. So, factor out the preprocessing/checking instead.

File: Orange/widgets/unsupervised/tests/test_owdistances.py [hunk 3]
```diff
@@ -62,3 +64,23 @@ def test_mahalanobis_error(self):
...
+        Users sees an error message when calculating too large arrays and Orange
+        does not crash.
+        GH-2315
+        """
+        self.assertEqual(len(self.widget.Error.active), 0)
+        self.send_signal("Data", self.iris)
+
+        mock = Mock()
+        mock.side_effect = ValueError
+        self.widget.compute_distances(mock, self.iris)
```
Comment (janezd) [inline Orange/widgets/unsupervised/tests/test_owdistances.py]: Use `side_effect` as a keyword argument for mock. I wouldn't even bother to have a `mock` variable here; I'd use `Mock(side_effect=ValueError)` as an argument.

File: Orange/widgets/unsupervised/tests/test_owdistances.py [hunk 4]
```diff
@@ -62,3 +64,23 @@ def test_mahalanobis_error(self):
...
+        self.assertTrue(self.widget.Error.distances_value_error.is_shown())
+
+        mock = Mock()
+        mock.side_effect = MemoryError
+        self.widget.compute_distances(mock, self.iris)
```
Comment (janezd) [inline Orange/widgets/unsupervised/tests/test_owdistances.py]: Same.

**Comment by jerneju** [comment]
Fixed.

**Comment by jerneju** [comment]
https://sentry.io/biolab/orange3/issues/204914018/

**Pull request merged**